fix: `OR` operator in where clauses should not be strict

msfstef opened the pull request:
Was trying to get a where clause in a shape to work, where either one field would match a value OR another field would, and in practice I would set one field to null and move the value over to the other (meaning that it should trigger an update) and I was seeing a delete coming through.

Digging through the code I found the issue to be that we were treating `AND`, `OR`, and `NOT` all as `strict?` operators, meaning that if any of their arguments are `NULL` we collapse the result to `NULL` also. However this is not the correct behaviour on postgres for the `OR` operator, as shown below:

```sql
select null and true; # -> null
select not null; # -> null
select null or true; # -> t
select null or false; # -> null
```

I've changed the `OR` operator to not be strict along with a simple test for it. I wonder if we should populate our where clause execution tests a bit more, we can probably generate a good suite of tests for it.


edit: I've added a custom `or` operator after @icehaunter 's note about the odd behaviour, and also had to fix up some "bool chains" we were parsing.

The affected operators are:
- `OR`  - before `true OR null` would be `null` rather than `true`
- `IN` - before `1 IN (1, null)` would be `null` rather than `true`, since `IN` is just a chain of `OR`s
- `BETWEEN SYMMETRIC` - no behaviour change as the underlying operations are comparisons that will always either both yield `null` or a boolean, but it now uses the correct `or` operator either way

icehaunter commented:
Not sure if worth correcting further, but `SELECT null or false` yields `null`, not `false` in postgres /shrug

msfstef commented:
> Not sure if worth correcting further, but SELECT null or false yields null, not false in postgres /shrug

Ugh I did not think of checking that case - that's disgusting. It's probably worth fixing as I'm worried it might lead to other unexpected issues like the one I had down the line. I guess I can special case the execution of the OR function, not great but this "semi-strict" behaviour is probably not worth encoding otherwise?

icehaunter commented:
@msfstef your test work with this argument order but not the other way around I believe

msfstef commented:
Fixed it up with a custom `pg_or` function, added more tests and included reversed arguments for safety.

Turns out the `IN` operator was also wrong, since it chains `OR`s and thus `1 IN (1, NULL)` was returning `NULL` rather then true.
